internal/testutils/pipe_listener.go

```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package testutils
+
+import (
+	"errors"
+	"net"
+	"time"
+)
+
+type pipeAddr struct{}
+
+func (p pipeAddr) Network() string { return "pipe" }
+func (p pipeAddr) String() string  { return "pipe" }
+
+// PipeListener is a listener with an unbuffered pipe. Each write will complete only once the other side reads.
+type PipeListener struct {
+	C    chan chan<- net.Conn
+	done chan struct{}
```

Comment:
Why was this chan added?  There are some problems (below) about determinism now that a second channel is involved.

If the issue is that the dialer panics once the listener is closed...I think that's OK - we just have to make sure we close the client before closing the listener, and without any dials in flight.  If we need this to work we should maybe talk about a different design.  We may be able to just add in a lock.

Reply:
> Why was this chan added?

To try to simulate the same behavior as net.Listener, which is that calling close unblocks all Accepts.

I've removed it and replaced it with a simple select in the dialer that causes an error to be returned if there's no corresponding Accept waiting for the dial.